codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -307,38 +338,87 @@ private Http2Stream requireStream(int streamId) {
+                        // write padding only
+                        int writeablePadding = Math.min(allowedFrameSize, padding);
+                        padding -= writeablePadding;
+                        bytesWritten += writeablePadding;
+                        ChannelPromise writePromise;
+                        if (size == bytesWritten && !promise.isVoid()) {
+                            // Can use the original promise if it's the last write
+                            writePromise = promise;
+                        } else {
+                            // Create a new promise and listen to it for failure
+                            writePromise = ctx.newPromise();
+                            writePromise.addListener(this);
+                        }
+                        frameWriter()
+                                .writeData(ctx, stream.id(), Unpooled.EMPTY_BUFFER,
+                                        writeablePadding, size == bytesWritten && endOfStream,
+                                        writePromise);
+                        return;
+                    }
+                    FileRegion toWrite;
+                    int writeableData = (int) data.count();
+                    if (writeableData > allowedFrameSize) {
+                        FileRegion[] regions = data.split(allowedFrameSize);
+                        toWrite = regions[0];
+                        data = regions[1];
+                    } else {
+                        toWrite = data;
+                        data = null;
+                    }
+                    int writeablePadding = Math.min(allowedFrameSize - writeableData, padding);
+                    padding -= writeablePadding;
+                    bytesWritten += writeableData + writeablePadding;
+                    ChannelPromise writePromise;
+                    if (size == bytesWritten && !promise.isVoid()) {
+                        // Can use the original promise if it's the last write
+                        writePromise = promise;
+                    } else {
+                        // Create a new promise and listen to it for failure
+                        writePromise = ctx.newPromise();
+                        writePromise.addListener(this);
+                    }
+                    frameWriter().writeData(ctx, stream.id(), toWrite, writeablePadding,
+                            size == bytesWritten && endOfStream, writePromise);
+                } while (size != bytesWritten && allowedBytes > bytesWritten);
+            } finally {
+                size -= bytesWritten;
+            }
         }
 
         @Override
-        public void error(Throwable cause) {
-            ReferenceCountUtil.safeRelease(data);
-            lifecycleManager.onException(ctx, cause);
-            data = null;
-            size = 0;
-            promise.tryFailure(cause);
+        public boolean merge(FlowControlled next) {
+            return false;
```

Comment:
If we do move forward with the PR lets create an issue for followup work to create a `CompositeFileRegion`.